src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -285,101 +307,126 @@ object RuleUtils {
-          if (useBucketSpec || !index.hasParquetAsSourceFormat || filesDeleted.nonEmpty ||
-              location.partitionSchema.nonEmpty) {
-            // Since the index data is in "parquet" format, we cannot read source files
-            // in formats other than "parquet" using one FileScan node as the operator requires
-            // files in one homogenous format. To address this, we need to read the appended
-            // source files using another FileScan node injected into the plan and subsequently
-            // merge the data into the index data. Please refer below [[Union]] operation.
-            // In case there are both deleted and appended files, we cannot handle the appended
-            // files along with deleted files as source files do not have the lineage column which
-            // is required for excluding the index data from deleted files.
-            // If the source relation is partitioned, we cannot read the appended files with the
-            // index data as the schema of partitioned files are not equivalent to the index data.
-            unhandledAppendedFiles = filesAppended
-            index.content.files
-          } else {
-            // If BucketSpec of index data isn't used (e.g., in the case of FilterIndex currently)
-            // and the source format is parquet, we could read the appended files along
-            // with the index data.
-            index.content.files ++ filesAppended
-          }
-        }
-
-        // In order to handle deleted files, read index data with the lineage column so that
-        // we could inject Filter-Not-In conditions on the lineage column to exclude the indexed
-        // rows from the deleted files.
-        val newSchema = StructType(
-          index.schema.filter(s =>
-            baseRelation.schema.contains(s) || (filesDeleted.nonEmpty && s.name.equals(
-              IndexConstants.DATA_FILE_NAME_ID))))
-
-        val newLocation = new InMemoryFileIndex(spark, filesToRead, Map(), None)
-        val relation = HadoopFsRelation(
-          newLocation,
-          new StructType(),
-          newSchema,
-          if (useBucketSpec) Some(index.bucketSpec) else None,
-          new ParquetFileFormat,
-          Map(IndexConstants.INDEX_RELATION_IDENTIFIER))(spark)
-
-        val updatedOutput =
-          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
-
-        if (filesDeleted.isEmpty) {
-          baseRelation.copy(relation = relation, output = updatedOutput)
-        } else {
-          val lineageAttr = AttributeReference(IndexConstants.DATA_FILE_NAME_ID, LongType)()
-          val deletedFileIds = filesDeleted.map(f => Literal(f.id)).toArray
-          val rel =
-            baseRelation.copy(relation = relation, output = updatedOutput ++ Seq(lineageAttr))
-          val filterForDeleted = Filter(Not(In(lineageAttr, deletedFileIds)), rel)
-          Project(updatedOutput, OptimizeIn(filterForDeleted))
-        }
```

Comment:
Extracted as the `getIndexHadoopFsRelation` method at [L364-L398](https://github.com/microsoft/hyperspace/pull/321/files#diff-8c312528287b51f613809d58a7b14b23c151d65794c099b653605795e0a738edR364-R398).